### ISSUE TITLE
web support

### ISSUE THREAD

**Issue by ayushin** [jonasbark/flutter_stripe_payment#139]
Hi guys!

I wonder what would be our options to add flutter web support?

**Comment by ayushin** [comment]
Looks like it would be possible to do this the same way https://github.com/Lyokone/flutterlocation is implemented for both mobile and web.

We could create flutter_stripe_payment_platform for platform abstraction and flutter_stripe_payment_web to use https://stripe.com/docs/js/

Any interest in this?

**Comment by peterjeffrey** [comment]
@ayushin Would love this! Do you know if there are any workarounds or packages that enable integrating Stripe on Flutter for Web?

**Comment by ayushin** [comment]
It seems there's a dart wrapper for stripe https://pub.dev/packages/stripe and also there's a fully native https://pub.dev/packages/stripe_sdk

Unfortunately I don't have much time to work on this now, but I would love to help where I can.

**Comment by ayushin** [comment]
good news!

1. Managed to generate functional js-interop library with [https://github.com/dart-lang/js_facade_gen](dart-js-facade-gen)

2. Got a few methods working

Will post prototype library soon.

If anybody else is working on this - let me know

**Comment by ayushin** [comment]
So I've made a [https://github.com/jonasbark/flutter_stripe_payment/pull/147](pr) initially for native pay but there are quite a few open questions:

- native pay supports both Apple and Android so canMakeNativePay requires country, currency and returns null if not supported or object with applePay = true for apple pay

- paymentRequestWithNativePay takes double, but stripe sdk expects int in smallest currency, for now hacked as total * 100 but that's not right

- initially i wanted to implement this as flutter_stripe_payment_web but because of the above this doesn't work without modifying the original plugin

- we should be using PlatformInterface to avoid platform specific methods i.e. Platform.isIOS etc inside of the plugins

- quite unfortunately the original plugin models do not coincide with stripe models generated from typescript (i.e Token, BankAccount)

All in all some higher level refactoring would be needed to have all 3 platforms integrated seamlessly in one interface (I don't really see the need for separate ApplePayOptions and AndroidPayOptions - stripe doesn't have them).

I'm very short on time, so if anybody wants to help - I would be super happy.

**Comment by gorillatapstudio** [comment]
any progress on this issue? Really looking forward to using this sdk for web payment. thanks!

**Comment by ayushin** [comment]
Well the PR above works for us so far, but it requires a bit of adaptation as per above.

**Comment by gorillatapstudio** [comment]
Thanks for working on that! Looking forward to a new release!

**Comment by eliudio** [comment]
"Well the PR above works for us so far, but it requires a bit of adaptation as per above."

I tried to find the PR, and seems gone. Is it correct to assume this has been pushed to master? If so, can we conclude stripe_payment supports flutterweb? If so, would be great to make that visible in pub.dev.

**Comment by ayushin** [comment]
There's this breaking change:

`native pay supports both Apple and Android so canMakeNativePay requires country, currency and returns null if not supported or object with applePay = true for apple pay
`

and the proper implementation requires refactoring the plugin... if a day had 72 hours.... :-)

**Comment by gorillatapstudio** [comment]
@ayushin are you still moving this forward in the near future? Thanks!

**Comment by ayushin** [comment]
@gorillatapstudio we use this in production:

```
  stripe_payment:
    git:
      url: https://github.com/apexlabs-ai/flutter_stripe_payment
      ref: web
```

The only plan we have is to split it into separate `flutter_stripe_payment_web` but on hold due to lack of resources atm.

**Comment by naikdp7** [comment]
The stripe_payment plugin for web doesn't implement the method 'createTokenWithCard' does createTokenWith card supported in the web branch?i have added  

```
stripe_payment:
    git:
      url: https://github.com/apexlabs-ai/flutter_stripe_payment
      ref: web
```

**Comment by ayushin** [comment]
that could be - i'll try to have a look at it when i have time, however it can be we are not using that functionality at all

**Comment by naikdp7** [comment]
i have check the code.there is no case for createTokenWithCard. though i am adding that you code.

**Comment by gorillatapstudio** [comment]
@ayushin I am using the apexlabs-ai branch and have two issues to report. Not sure if it only happened to me or everyone.

(1) when i tried to call 

> StripePayment.paymentRequestWithCardForm(CardFormPaymentRequest());
it showed an error of 

> The stripe_payment plugin for web doesn't implement the method 'paymentRequestWithCardForm'

(2) when the payment intent succeed (confirmed in stripe), paymentIntentResult.status is null (although I do receive "succeeded" when running on ios or android native app).

Everything else is working great! Thanks!

**Comment by ayushin** [comment]
1. See the above:

> native pay supports both Apple and Android so canMakeNativePay requires country, currency and returns *null* if not supported or object with applePay = true for apple pay

2. Did not get to implement this yet, should be very easy, but in our case we pass credit card details from own flutter form

**Comment by gorillatapstudio** [comment]
Thanks @ayushin ! so for web stripe payment, would you further clarify what I should do to make sure the paymentIntentResult.status  can return either "succeeded" or "failed" but not null?  All my code for web and native payment are the same at the moment.

**Comment by ayushin** [comment]
Well, there's no separate method for "not supported", so that's why you get null to indicate that. We'd need a different interface all together if we don't want null there (like an enum)

**Comment by gorillatapstudio** [comment]
However, my payment was charged successfully at stripe. Does it mean the status variable in callback was not supported for web? In the meantime, is there anyway I can do to know if the payment was successful or not. Thanks!

**Comment by yaizudamashii** [comment]
> @ayushin I am using the apexlabs-ai branch and have two issues to report. Not sure if it only happened to me or everyone.
> 
> (1) when i tried to call
> 
> > StripePayment.paymentRequestWithCardForm(CardFormPaymentRequest());
> 
> it showed an error of
> 
> > The stripe_payment plugin for web doesn't implement the method 'paymentRequestWithCardForm'
> 
> (2) when the payment intent succeed (confirmed in stripe), paymentIntentResult.status is null (although I do receive "succeeded" when running on ios or android native app).
> 
> Everything else is working great! Thanks!

@gorillatapstudio I fail to see what is working, if it is working at all.  None of the methods in the example works for web, namely
```
createSourceWithParams
paymentRequestWithCardForm
createTokenWithCard
createPaymentMethod
paymentRequestWithNativePay
```
Any one of them returns error of either not configured or is available on web.  Did you test on web?  What did you do to make anything work?

**Comment by ayushin** [comment]
@yaizudamashii we do payments via backend so we did not have need for those methods... sorry for half-ready solution

paymentRequestWithNativePay should work though! it works for us on web

**Comment by yaizudamashii** [comment]
I see that canMakeNativePayPayments is returning null, so somehow the browser does not allow payment.  Maybe there is some problem with authentication?

**Comment by yaizudamashii** [comment]
@ayushin I read the code a bit more, is this web extension designed to be used in a browser in iOS devices?  Do I understand correctly that `canMakeNativePayPayments` is true if there is Apple Pay available?

**Comment by gorillatapstudio** [comment]
@yaizudamashii use the http call to your backend server for createPaymentIntent. However the StripePayment.confirmPaymentIntent is partially implemented for web I guess since it succeed in stripe dashboard but didn’t return status code.

**Comment by ayushin** [comment]
Yes, null means it it not supported, make sure you try it with `https`

**Comment by yaizudamashii** [comment]
Ok I read the Stripe documentation, no wonder I was getting null back, even when trying with ngrok for https

`paymentRequest.canMakePayment only returns true on Safari 10.1+ (desktop and mobile), Chrome 61+ (desktop and mobile), and Microsoft Edge 16.16299+ when the customer has a card already saved with Apple Pay, Microsoft Pay, or their browser.`
[https://stripe.com/docs/js/payment_request/can_make_payment](https://stripe.com/docs/js/payment_request/can_make_payment)

**Comment by yaizudamashii** [comment]
Thank you @ayushin Finally I was able to receive the payment token using chrome with credit card saved, with https, localhost forwarded via ngrok.
The example is still not perfect, as the window did not disappear after receiving the token and prevented me from clicking on "Complete Native Payment" button.  I might look into how to dismiss that window.

**Comment by ayushin** [comment]
you need to call completeNativePayment method (or similar - long time since i saw the code) for apple pay

**Comment by FallenRiteMonk** [comment]
any update on this? i would need this for a upcoming project.

**Comment by dgriff24** [comment]
Based on [the README file](https://github.com/apexlabs-ai/flutter_stripe_payment/blob/web/README.md), it appears that you can only make use of the Stripe API on Android or iOS devices. Can I, in its current form, use this package for websites on Windows or Mac devices?

**Comment by jonasbark** [comment]
@dgriff24 it really only works on Android and iOS.
If you need the API only (vs UI etc), then there are other packages you can use.
